src/main/java/com/faforever/client/map/generator/MapGeneratorService.java

```diff
@@ -107,13 +119,40 @@ private void deleteGeneratedMaps() {
     }
   }
 
-
   public CompletableFuture<String> generateMap() {
-    return generateMap(queryNewestVersion(), seedGenerator.nextLong());
+    ByteBuffer seedBuffer = ByteBuffer.allocate(8);
+    seedBuffer.putLong(seedGenerator.nextLong());
+    String seedString = Base64.getEncoder().encodeToString(seedBuffer.array());
+    return generateMap(generatorVersion,seedString);
+  }
+
+  public CompletableFuture<String> generateMap(byte[] optionArray) {
+    ByteBuffer seedBuffer = ByteBuffer.allocate(8);
+    seedBuffer.putLong(seedGenerator.nextLong());
+    String seedString = Base64.getEncoder().encodeToString(seedBuffer.array());
+    String optionString = Base64.getEncoder().encodeToString(optionArray);
+    return generateMap(generatorVersion,seedString+'_'+optionString);
+  }
+
+  public CompletableFuture<String> generateMap(String version, byte[] optionArray) {
+    ByteBuffer seedBuffer = ByteBuffer.allocate(8);
+    seedBuffer.putLong(seedGenerator.nextLong());
+    String seedString = Base64.getEncoder().encodeToString(seedBuffer.array());
+    String optionString = Base64.getEncoder().encodeToString(optionArray);
+    return generateMap(version,seedString+'_'+optionString);
+  }
+
+  public CompletableFuture<String> generateMap(ComparableVersion version, byte[] optionArray) {
+    ByteBuffer seedBuffer = ByteBuffer.allocate(8);
+    seedBuffer.putLong(seedGenerator.nextLong());
+    String seedString = Base64.getEncoder().encodeToString(seedBuffer.array());
+    String optionString = Base64.getEncoder().encodeToString(optionArray);
+    return generateMap(version,seedString+'_'+optionString);
   }
 
   @VisibleForTesting
-  protected String queryNewestVersion() {
+  @Cacheable(CacheNames.MAP_GENERATOR)
```

Comment:
Make sue this is only ever called via a Spring Proxy

Reply:
Should only be called by the bean